DOC clarifications on the release process (#15759)
scikit-learn/scikit-learn: doc/developers/maintainer.rst

Maintainer / core-developer information
========================================

A release is either a major release, or a minor bug fix release. Our convention
is that we release one or more release candidates (0.XXrcN) before releasing

> the final binaries. Creating the first RC involves creating the release branch

Reviewer: ```suggestion
the final distributions. Creating the first RC involves creating the release branch
```

since we also ship source dists?

branch (0.99.X for instance) directly on the main repo. The rest of the changes
should be done as a PR to the master branch, if it's possible, or done through
a PR to the release branch, for the other maintainers to review before the

> release. This is further explained under *Preparing a release PR*.

Reviewer 2: @adrinjalali one may expect a link to the corresponding subsection here? 
```suggestion
release. This is further explained under :ref:`Preparing a release PR<Preparing a release PR>`.
```

Maintainer / core-developer information
========================================

A release is either a major release, or a minor bug fix release. Our convention
is that we release one or more release candidates (0.XXrcN) before releasing
the final distributions. Creating the first RC involves creating the release
branch (0.99.X for instance) directly on the main repo. The rest of the changes
should be done as a PR to the master branch, if it's possible, or done through
a PR to the release branch, for the other maintainers to review before the
release. This is further explained under

> :ref:`Preparing a release PR<Preparing a release PR>`. We follow the

Reviewer 2: ... me asking for forgiveness for the sphinx warning... this time I've checked
 ```suggestion
`Preparing a release PR<#Preparing-a-release-PR>`_. We follow the
```

Author: haha, thanks for the fix.

Maintainer / core-developer information
========================================

A major release, incrementing the minor version, or a big fix release

> incrementing the patch version. Our convention is that we release one or more

Reviewer: that sentence is missing something

also big -> bug?

release candidates (0.XXrcN) before releasing the final distributions. Creating

> the first RC involves creating the release branch (0.99.X for instance)

Reviewer: using X for 2 different things here?

> directly on the main repo. The rest of the changes should be done as a PR to

Reviewer 3: Could we call this stage "Freeze". Supposidely, we should not add API change, new feature after this stage.

> the master branch, if it's possible, or done through a PR to the release

Reviewer: what are "the rest of the changes?" in this context?

branch, for the other maintainers to review before the release. Note that all
development for the minor releases and the major release of for instance 0.99,
happens under the branch called ``0.99.X``, where ``X`` is a literal, and
independent of minor bug fix releases. Each release candidate, major, or minor
release is a tag under the same branch. This is further explained under

> `Preparing a release PR<#Preparing-a-release-PR>`_. We follow the `PEP101

Reviewer 2: shame on me....
line 15 should be
```
:ref: `preparing-a-release-pr`. We follow the `PEP101`
```
Then a label should be added before the section:
```
.. _preparing-a-release-pr:
```
:(

Author: no worries @cmarmo, I'm actually quite happy you figured it out. I'm not the most fluent when it comes to ReST as you can see :P

Maintainer / core-developer information
========================================

A major release, incrementing the minor version, or a bug fix release
incrementing the patch version. Our convention is that we release one or more
release candidates (0.RRrcN) before releasing the final distributions. Creating
the first RC involves creating the release branch (0.99.X for instance)
directly on the main repo. This is considered the *feature freeze*. The rest of
the changes (required to fix some of the issues detected during the release
process are a good example) should be done as a PR to the master branch, if
it's possible, or done through a PR to the release branch, for the other
maintainers to review before the release. Note that all development for the
minor releases and the major release of for instance 0.99, happens under the
branch called ``0.99.X``, where ``X`` is a literal, and independent of minor

> bug fix releases. Each release candidate, major, or minor release is a tag

Reviewer: I don't understand this sentence, especially since it is said above that changes should be done to master whenever possible

Author: The development happens on that branch, through PRs which are merged into that branch.

> independent of minor bug fix releases. Each release candidate, major, or minor

Reviewer: I still think this is ambiguous given the previous sentence. Please clarify what "development happens" means or change the phrasing?

   change log is reasonably well curated. Some tools for these tasks include:

   - ``maint_tools/sort_whats_new.py`` can put what's new entries into
     sections. It's not perfect, and requires manual checking of the changes.
     If the whats new list if well curated, it may not be necessary.

   - The ``maint_tools/whats_missing.sh`` script may be used to identify pull
     requests that were merged but likely missing from What's New.

Preparing a release PR
......................

Since any commits to an existing branch (e.g. 0.999.X) will automatically
update the web site documentation, it is best to develop a release with a pull
request in which 0.999.X is the base. It also allows you to keep track of any

> tasks towards release with a TO DO list.

Reviewer: I don't understand this, what documentation is updated exactly, and why is it a bad thing?

Author: does this help?

Most development of the release, and its documentation, should happen in master
to avoid asynchrony. To select commits from master for use in the bug fix
(version 0.999.3), you can use::

    $ # # assuming master and upstream/master are the same
    $ git checkout -b release-0.999.3 master
    $ git rebase -i upstream/0.999.X

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others

> understand what's included.

Reviewer: Also explain what should and what should not be included during the rebase? My understanding is that:

- to create e.g. 0.21.rc0, we include everything from 0.20.X 
- to create e.g. 0.21.rc1, we only select specific commits from 0.21.rc0. I.e. we include additional bug fixes and minor docs but typically ignore new features that weren't already in rc0.

   change log is reasonably well curated. Some tools for these tasks include:

   - ``maint_tools/sort_whats_new.py`` can put what's new entries into
     sections. It's not perfect, and requires manual checking of the changes.
     If the whats new list if well curated, it may not be necessary.

   - The ``maint_tools/whats_missing.sh`` script may be used to identify pull
     requests that were merged but likely missing from What's New.

Permissions
...........

The release manager requires a set of permissions on top of the usual
permissions given to maintainers, which includes:

- *maintainer* role on ``scikit-learn`` projects on ``pypi.org`` and
  ``test.pypi.org``, separately.
- become a member of the *scikit-learn* team on conda-forge.
- *maintainer* on ``https://github.com/MacPython/scikit-learn-wheels``

Preparing a release PR
......................

Since any commits to an existing branch (e.g. 0.999.X) will automatically

> update the web site documentation, it is best to develop a release with a pull

Reviewer: Is it true that any branch to the repo has an online documentation? I thought we'd have a way to curate which branches we want to have the docs for, like with ReadTheDocs

Author: AFAIK any `0.XX.X` branch has a documentation online. Any change to the `0.21.X` branch for example, will be there almost immediately. And to add a new release, you just create the branch, and it'll automatically be there.

> request in which 0.999.X is the base. It also allows you to keep track of any

Reviewer: Not sure I understand what "0.999.X is the base" means.  does it mean that the PR tries to merge things into the 0.999.X branch? Is there any other way to do it anyway?

Author: For me the two git commands a few lines later explain what it is. I also didn't understand what it meant from the text.

     If the whats new list if well curated, it may not be necessary.

   - The ``maint_tools/whats_missing.sh`` script may be used to identify pull
     requests that were merged but likely missing from What's New.

Permissions
...........

The release manager requires a set of permissions on top of the usual
permissions given to maintainers, which includes:

- *maintainer* role on ``scikit-learn`` projects on ``pypi.org`` and
  ``test.pypi.org``, separately.
- become a member of the *scikit-learn* team on conda-forge.
- *maintainer* on ``https://github.com/MacPython/scikit-learn-wheels``

Preparing a release PR
......................

Since any commits to an existing branch (e.g. 0.999.X) will automatically
update the web site documentation, it is best to develop a release with a pull
request in which 0.999.X is the base. It also allows you to keep track of any
tasks towards release with a TO DO list. Since the documentation for the branch
is live as a stable release, it should ideally match the distributions
available to users. Having all the changes for a release in one PR minimizes
the gap between the time when the website points to a release, and the release

> being available to users.

Author: yes, having one PR just makes it clearer what the difference between two releases are.

> being available to users.

Reviewer: So the current description should be updated. Right now it says have only one PR helps minimizing the time gap, which isn't true

Most development of the release, and its documentation, should happen in master
to avoid asynchrony. To select commits from master for use in the bug fix
(version 0.999.3), you can use::

    $ # # assuming master and upstream/master are the same
    $ git checkout -b release-0.999.3 master

>     $ git rebase -i upstream/0.999.X

Reviewer: what branch name? the branch we're creating here is named release-0.999.3

branch `0.99.X` directly on the main repo, where `X` really is the letter X,
**not a placeholder**. This is considered the *feature freeze*. The
development for the major and minor releases of 0.99 should
**also** happen under `0.99.X`. Each release (rc, major, or minor) is a tag
under that branch.

In terms of including changes, the first RC ideally counts as a *feature
freeze*. Each coming release candidate and the final release afterwards will
include minor documentation changes and bug fixes. Any major enhancement or
feature should be excluded.

The minor releases should include bug fixes and some relevant documentation
changes only. Any PR resulting in a behavior change which is not a bug fix
should be excluded.

First, create a branch, **on your own fork** (to release e.g. `0.999.3`)::

    $ # assuming master and upstream/master are the same
    $ git checkout -b release-0.999.3 master

Then, create a PR **to the** `scikit-learn/0.999.X` **branch** (not to
master!) with all the desired changes::

	$ git rebase -i upstream/0.999.2

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
understand what's included.

Making a release

> ................

Reviewer: At what point do we merge the PR mentioned above into the 0.99.X branch? Should we add a step in the list below?

Author: It has to happen before tagging. You can add one if you think it's not clear.

   change log is reasonably well curated. Some tools for these tasks include:

   - ``maint_tools/sort_whats_new.py`` can put what's new entries into
     sections. It's not perfect, and requires manual checking of the changes.

>      If the whats new list if well curated, it may not be necessary.

Reviewer: ```suggestion
     If the whats new list is well curated, it may not be necessary.
```

     requests that were merged but likely missing from What's New.

Permissions
...........

The release manager requires a set of permissions on top of the usual
permissions given to maintainers, which includes:

- *maintainer* role on ``scikit-learn`` projects on ``pypi.org`` and
  ``test.pypi.org``, separately.
- become a member of the *scikit-learn* team on conda-forge.
- *maintainer* on ``https://github.com/MacPython/scikit-learn-wheels``

.. _preparing-a-release-pr:

Preparing a release PR
......................

Since any commits to an existing branch (e.g. 0.999.X) will automatically
update the web site documentation, it is best to develop a release with a pull
request in which 0.999.X is the base. It also allows you to keep track of any
tasks towards release with a TO DO list. Since the documentation for the branch
is updated live, it should ideally match the version available to users. Having
all the changes for a release in one PR minimizes the gap between the time when
the website points to a release, and the release being available to users.

Most development of the release, and its documentation, should happen in master

> to avoid asynchrony. First, create a branch, **on your own fork** (to release

Reviewer: This whole  first paragraph and the first sentence of the second one feel redundant with the rest

Author: Not sure how to fix that.

Reviewer: I have suggested an alternative paragraph before

Permissions
...........

The release manager requires a set of permissions on top of the usual
permissions given to maintainers, which includes:

- *maintainer* role on ``scikit-learn`` projects on ``pypi.org`` and
  ``test.pypi.org``, separately.
- become a member of the *scikit-learn* team on conda-forge.
- *maintainer* on ``https://github.com/MacPython/scikit-learn-wheels``

.. _preparing-a-release-pr:

Preparing a release PR
......................

Since any commits to an existing branch (e.g. 0.999.X) will automatically
update the web site documentation, it is best to develop a release with a pull
request in which 0.999.X is the base. It also allows you to keep track of any
tasks towards release with a TO DO list. Since the documentation for the branch
is updated live, it should ideally match the version available to users. Having
all the changes for a release in one PR minimizes the gap between the time when
the website points to a release, and the release being available to users.

Most development of the release, and its documentation, should happen in master
to avoid asynchrony. First, create a branch, **on your own fork** (to release
e.g. `0.999.3`)::

>     $ # assuming master and upstream/master are the same

Reviewer 4: For me `origin/master` is my fork, `upstream/master` is official scikit-learn master. I guess that's not what this means? I know different people have different conventions. Maybe better to avoid talking about `upstream` to avoid confusion if so. Not that in github docs `upstream` is also not one's fork https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/syncing-a-fork

Author: yes, I exactly meant `upstream` with the same semantics as you mention.

    $ git checkout -b release-0.999.3 master

Note that pushing the (empty) branch **to the main repo** will automatically
upload its documentation, which currently corresponds to that of master, so
it's not an accurate representation of what the release will actually be.

Then create a PR **to the `scikit-learn/0.999.X` branch and not `master`** with
all the desired changes::

> 	$ git rebase -i upstream/0.999.2

Reviewer 4: So this is this a branch in the fork or in the official scikit-learn repo? The documentation doesn't say how it was created..

Author: I added a step 0

The rest of the process is to merge the PR, create a tag and publish the
wheels. Ideally, this should be done quickly in order to minimize the gap
between the time when the website points to a release, and the release being
available to users.

It's nice to have a copy of the ``git rebase -i`` log in the PR to help others
understand what's included.

In terms of including changes, the first RC is a pure copy of the ``master``

> with a release version number, and ideally counts as a *feature freeze*. Each

Reviewer 4: I don't think it's a helpful technical detail. One can create the release branch, try to build wheels, see something that fails, add commits on top to fix it etc.  Saying that the first RC should be the exact copy of master is not true in general.

   - Update the release date in whats_new.rst

   - Edit the doc/templates/index.html to change the 'News' entry of the front page.

   - Note that these changes should be made in master and cherry-picked into
     the release branch, only before the final release.

2. On the branch for releasing, update the version number in
   sklearn/__init__.py, the ``__version__`` variable by removing ``dev*`` only
   when ready to release. On master, increment the version in the same place
   (when branching for release). This means while we're in the release
   candidate period, the latest stable is two versions behind the master

>    branch, and one one.

Reviewer 5: One one was a racehorse two two was one two. One one one one race. Two two one one two. 

Not sure what one one is doing here.

       $ python setup.py sdist

   - You can also test a binary dist build using::

       $ python setup.py bdist_wheel

   - You can test if PyPi is going to accept the package using::

       $ twine check dist/*

   You can run ``twine check`` after step 5 (fetching artifacts) as well.

   The result should be in the `dist/` folder. We will upload it later
   with the wheels. Check that you can install it in a new virtualenv and
   that the tests pass.

4. Proceed with caution. Ideally, tags should be created when you're almost

>    certain that the release is ready. You can test upload the sdist to

Reviewer 5: To avoid premature tagging,...

>    test.pypi.org, and follow step 5 and have ``BUILD_COMMIT`` set to the

Reviewer 5: ```suggestion
   test.pypi.org, and test the next step by setting ``BUILD_COMMIT`` to the
```

>    branch name (``0.22.X`` for instance), and test the wheel builds on a PR on

Reviewer 5: ```suggestion
   branch name (``0.22.X`` for instance) in a PR to the wheel building repo.

       $ python setup.py sdist

   - You can also test a binary dist build using::

       $ python setup.py bdist_wheel

   - You can test if PyPi is going to accept the package using::

       $ twine check dist/*

   You can run ``twine check`` after step 5 (fetching artifacts) as well.

   The result should be in the `dist/` folder. We will upload it later
   with the wheels. Check that you can install it in a new virtualenv and
   that the tests pass.

4. Proceed with caution. Ideally, tags should be created when you're almost
   certain that the release is ready, since adding a tag to the main repo can
   trigger certain automated processes. You can test upload the ``sdist`` to
   ``test.pypi.org``, and test the next step by setting ``BUILD_COMMIT`` to the
   branch name (``0.22.X`` for instance) in a PR to the wheel building repo.
   Once all works, you can proceed with tagging. Create the tag and push it (if

>    it's an RC, it can be ``0.xxrc1`` for instance)::

Reviewer: I understand the tagging should be done in the 0.99.X branch?

Author: yes, correct.

       $ python setup.py sdist

   - You can also test a binary dist build using::

       $ python setup.py bdist_wheel

   - You can test if PyPi is going to accept the package using::

>        $ twine check dist/*

Reviewer 4: Not sure about the point of building the wheel locally and running `twine check` on it. I think using only wheels built in CI is less prone to confusion.

Author: We still have to build the `sdist` locally anyway. So we should check that one at least locally. Also, the build CI is pretty slow, and doing a check before triggering that is not a bad idea.

>    You can run ``twine check`` after step 5 (fetching artifacts) as well.

Reviewer 4: Let's move it there instead then.

Author: I put it here cause step 5 is after creating the tag. Ideally once the tag is created, there shouldn't be more necessary changes.

   The result should be in the `dist/` folder. We will upload it later
   with the wheels. Check that you can install it in a new virtualenv and
   that the tests pass.

4. Proceed with caution. Ideally, tags should be created when you're almost

>    certain that the release is ready to avoid premature tagging. You can test

Reviewer 4: ```suggestion
   certain that the release is ready. You can test
```
as it's redundant otherwise..

   upload the sdist to test.pypi.org, and test the next step by setting
   ``BUILD_COMMIT`` to the branch name (``0.22.X`` for instance) in a PR to the
   wheel building repo. Once all works, you can proceed with tagging. Create
   the tag and push it (if it's an RC, it can be 0.xxrc1 for instance)::

    $ git tag -a 0.999

>     $ git push git@github.com:scikit-learn/scikit-learn.git --tags

Reviewer 4: No -- you can have multiple unrelated tags locally.
```suggestion
    $ git push git@github.com:scikit-learn/scikit-learn.git 0.999
```

Author: I actually do steps 3-6 on a different clone where `origin` is the main repo, and is clean of all branches tags of my own fork, should I put that in?

Reviewer 4: I pushed a commit to explicitly indicate the pushed tag. I would not be comfortable using the above command as there is always a risk of making an error on the repo in one is etc.

       $ git checkout master
       $ rm stable
       $ ln -s 0.999 stable

>        $ sed -i "s/latestStable = '.*/latestStable = '0.999';/g" versionwarning.js

Reviewer 5: is g necessary?

Author: yeah, my `sed` complained about an unfinished `s/` directive or something w/o it.

Reviewer 5: You're right. It needs the / but not the g

>        $ git commit -am "Update stable to point to 0.999"

Reviewer 5: I don't get why you prefer this to explicitly only committing stable

Author: the `stable` used to be the only thing changing. Now `versionwarning.js` also changes in this step.

Reviewer 4: Added `git add stable/ versionwarning` on the line above to be explicit.
